[VPlan] Sink recipes from the vector loop region in licm.

Mel-Chen opened the pull request:
When a recipe can be safely sunk and all of its users are outside the vector loop region in the same dedicated exit block, the recipe does not need to be executed on every iteration.
This patch extends the VPlan-based LICM (Loop Invariant Code Motion) to also sink such recipes from the vector loop region into the exit block. This reduces redundant computation and improves cost model accuracy.

TODO: Support nested loop sinking
TODO: Support sinking `VPReplicateRecipe` (requires `replicateByVF` fixes)
TODO: Support recipes with multiple defined values (e.g., interleaved loads)
TODO: Clone recipes without users to all exit blocks
TODO: Support PHI node users by checking incoming value blocks
TODO: Support sinking when users are in multiple blocks
TODO: Clone recipes when users are on multiple exit paths

Co-authored-by: Luke Lau <luke@igalia.com>

Mel-Chen committed:
sink recipes from loop region

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2224,16 +2224,46 @@ void VPlanTransforms::cse(VPlan &Plan) {
+      if (any_of(Def->users(), [LoopRegion](VPUser *U) {
+            auto *UR = cast<VPRecipeBase>(U);
+            return UR->getParent()->getEnclosingLoopRegion();
+          }))
```
Did you mean to check enclosing loop region against LoopRegion?

Mel-Chen replied:
Updated the check. 73f354dea800905c966f9a12be7a98a0e04fba7c
I want to check that all users are outside the loop where the recipe is defined. 
The new version handles nested loops by starting to sink from the innermost loop.

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2224,16 +2224,46 @@ void VPlanTransforms::cse(VPlan &Plan) {
+      if (isDeadRecipe(R)) {
+        R.eraseFromParent();
+        continue;
+      }
```
Would be good to get removeRedundantRecipes to handle this?

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2250,6 +2280,13 @@ static void licm(VPlan &Plan) {
+static void licm(VPlan &Plan) {
+  VPRegionBlock *LoopRegion = Plan.getVectorLoopRegion();
+  sinkRecipesFromLoopRegion(LoopRegion);
+  hoistRecipesFromLoopRegion(LoopRegion);
```
```suggestion
  licmSink(Plan);
  licmHoist(Plan);
```

Not sure if this function is necessary: maybe inline in optimize?

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2224,16 +2224,46 @@ void VPlanTransforms::cse(VPlan &Plan) {
-static void licm(VPlan &Plan) {
-  VPBasicBlock *Preheader = Plan.getVectorPreheader();
+/// Sink recipes with no users inside the vector loop region into a dedicated
+/// exit block.
```
```suggestion
/// Sink loop-invariant recipes in the vector loop region into the middle block.
```

Mel-Chen committed:
Fix deletion of VPValue with users after sinking replicate recipes

VPlanTransforms::replicateByVF visits blocks outside the vector region
before those inside it. After sinking replicate recipes to the exit block,
the visitation order changes from unpack --> replicate` to replicate -->
unpack. This causes unpack to still have extractelement users when being
deleted.

This patch replaces the extractelement generated by cloned replicate
recipes with the corresponding extractelement from the unpack recipe.

Mel-Chen committed:
Sink innermost first

Mel-Chen commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2224,16 +2224,46 @@ void VPlanTransforms::cse(VPlan &Plan) {
+
+  for (VPBasicBlock *VPBB : VPBlockUtils::blocksOnly<VPBasicBlock>(
+           vp_depth_first_shallow(LoopRegion->getEntry()))) {
+    for (VPRecipeBase &R : make_early_inc_range(reverse(*VPBB))) {
```
Could you give me an example to show we have to checking that the operands are defined outside loop regions?

Mel-Chen commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2224,16 +2224,46 @@ void VPlanTransforms::cse(VPlan &Plan) {
+  VPBasicBlock *Preheader = LoopRegion->getPlan()->getVectorPreheader();
 
   // Hoist any loop invariant recipes from the vector loop region to the
   // preheader. Preform a shallow traversal of the vector loop region, to
```
Remove the hoistRecipesFromLoopRegion so that we don't need to move it, thanks.

Mel-Chen committed:
Fix the cost model issue

Mel-Chen committed:
Change traversal way

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2376,14 +2376,66 @@ void VPlanTransforms::cse(VPlan &Plan) {
+        if (cannotHoistOrSinkRecipe(R))
+          continue;
+
+        auto *Def = cast<VPSingleDefRecipe>(&R);
```
Why is this cast guaranteed not to assert?

Mel-Chen committed:
remove Visited

Mel-Chen committed:
Revert "Fix deletion of VPValue with users after sinking replicate recipes"

This reverts commit 67e8199f0cb7a0e5ad05e6528153bf839af4162e.

Mel-Chen committed:
don't sink llvm.assume

Mel-Chen commented:
Compared to #172343, the main difference is NCD-based LICM versus dedicated-exit-based LICM. I’m not sure whether there is any practical difference at the moment. One initial question worth considering is whether, in the presence of multiple exit blocks, an NCD could end up being inside the loop.

Aside from that, the differences are mostly in safety checks.

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2496,14 +2496,63 @@ void VPlanTransforms::cse(VPlan &Plan) {
+        if (any_of(Def->users(), [&](VPUser *U) {
+              VPRegionBlock *ParentL =
+                  cast<VPRecipeBase>(U)->getParent()->getEnclosingLoopRegion();
+              return ParentL && !is_contained(WorklistForSink, ParentL);
```
Could make WorklistForSink a SmallSetVector, and use ::contains?

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2496,14 +2496,63 @@ void VPlanTransforms::cse(VPlan &Plan) {
+          R.eraseFromParent();
+          continue;
+        }
+        if (cannotHoistOrSinkRecipe(R) || R.mayHaveSideEffects())
```
Doesn't cannot-hoist-or-sink check side-effects?

Mel-Chen replied:
Maybe we could split `cannotHoistOrSinkRecipe` into `cannotHoistRecipe` and `cannotSinkRecipe`?

Mel-Chen committed:
use SmallSetVector to record loops

Mel-Chen committed:
sink recipe after hoist

Mel-Chen committed:
Regress LICM sink to only handle outermost vector loop region

Nested loop sinking will be re-added in a follow-up patch if needed.

Mel-Chen committed:
remove dead recipe

Mel-Chen committed:
Don't sink if no user

Co-authored-by: Luke Lau <luke@igalia.com>

Mel-Chen committed:
Support multile exits

Mel-Chen committed:
use NCD-based sinking

Co-authored-by: Luke Lau <luke@igalia.com>

Mel-Chen commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2496,14 +2496,63 @@ void VPlanTransforms::cse(VPlan &Plan) {
 /// Move loop-invariant recipes out of the vector loop region in \p Plan.
 static void licm(VPlan &Plan) {
-  VPBasicBlock *Preheader = Plan.getVectorPreheader();
+  VPRegionBlock *LoopRegion = Plan.getVectorLoopRegion();
```
2d4ed65cadf3396aa7c6d80ec3adc7ecbcc9b7fb
Looking at the diff, deciding whether to hoist or sink first is quite an art.

Mel-Chen commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2496,14 +2496,63 @@ void VPlanTransforms::cse(VPlan &Plan) {
+    // TODO: Should check all predecessors of the exit block.
+    if (!SingleExit || SingleExit->getSinglePredecessor() != CurLoop)
+      continue;
+
```
> Thanks for taking a look at this. For my use case though the first-active-lane recipe was only used in the vector.early.exit block so it didn't require sinking into multiple blocks, at least from what I saw.
> 
> Did the diff you attached come from a particular test case?

This example is taken from `llvm/test/Transforms/LoopVectorize/RISCV/early-exit.ll` in #172454. The full context is:
```
VPlan 'Initial VPlan for VF={vscale x 1,vscale x 2,vscale x 4,vscale x 8,vscale x 16},UF>=1' {
Live-in vp<%0> = VF
Live-in vp<%1> = VF * UF
Live-in vp<%2> = vector-trip-count
Live-in ir<64> = original trip-count

ir-bb<entry>:
  IR   %p1 = alloca [1024 x i8], align 1
  IR   %p2 = alloca [1024 x i8], align 1
  IR   call void @init_mem(ptr %p1, i64 1024)
  IR   call void @init_mem(ptr %p2, i64 1024)
Successor(s): scalar.ph, vector.ph

vector.ph:
  vp<%3> = DERIVED-IV ir<3> + vp<%2> * ir<1>
  EMIT vp<%4> = step-vector i64
Successor(s): vector loop

<x1> vector loop: {
  vector.body:
    EMIT vp<%5> = CANONICAL-INDUCTION ir<0>, vp<%index.next>
    vp<%6> = DERIVED-IV ir<3> + vp<%5> * ir<1>
    vp<%7> = SCALAR-STEPS vp<%6>, ir<1>, vp<%0>
    CLONE ir<%arrayidx> = getelementptr inbounds ir<%p1>, vp<%7>
    vp<%8> = vector-pointer inbounds ir<%arrayidx>
    WIDEN ir<%ld1> = load vp<%8>
    CLONE ir<%arrayidx1> = getelementptr inbounds ir<%p2>, vp<%7>
    vp<%9> = vector-pointer inbounds ir<%arrayidx1>
    WIDEN ir<%ld2> = load vp<%9>
    WIDEN ir<%cmp3> = icmp ne ir<%ld1>, ir<%ld2>
    EMIT vp<%first.active.lane> = first-active-lane ir<%cmp3>
    EMIT vp<%10> = icmp ult vp<%4>, vp<%first.active.lane>
    CLONE ir<%arrayidx2> = getelementptr ir<%p0>, vp<%7>
    vp<%11> = vector-pointer ir<%arrayidx2>
    WIDEN ir<%ld3> = load vp<%11>, vp<%10>
    EMIT vp<%index.next> = add nuw vp<%5>, vp<%1>
    EMIT vp<%12> = any-of ir<%cmp3>
    EMIT vp<%13> = icmp eq vp<%index.next>, vp<%2>
    EMIT branch-on-two-conds vp<%12>, vp<%13>
  No successors
}
Successor(s): vector.early.exit, middle.block

middle.block:
  EMIT vp<%15> = extract-last-part ir<%ld3>
  EMIT vp<%16> = extract-last-lane vp<%15>
  EMIT vp<%cmp.n> = icmp eq ir<64>, vp<%2>
  EMIT branch-on-cond vp<%cmp.n>
Successor(s): ir-bb<loop.end>, scalar.ph

vector.early.exit:
  EMIT vp<%early.exit.value> = extract-lane vp<%first.active.lane>, ir<%ld1>
Successor(s): ir-bb<loop.end>

ir-bb<loop.end>:
  IR   %retval = phi i8 [ %ld1, %loop ], [ %ld3, %loop.inc ] (extra operands: vp<%16> from middle.block, vp<%early.exit.value> from vector.early.exit)
No successors

scalar.ph:
  EMIT-SCALAR vp<%bc.resume.val> = phi [ vp<%3>, middle.block ], [ ir<3>, ir-bb<entry> ]
Successor(s): ir-bb<loop>

ir-bb<loop>:
  IR   %index = phi i64 [ %index.next, %loop.inc ], [ 3, %entry ] (extra operand: vp<%bc.resume.val> from scalar.ph)
  IR   %arrayidx = getelementptr inbounds i8, ptr %p1, i64 %index
  IR   %ld1 = load i8, ptr %arrayidx, align 1
  IR   %arrayidx1 = getelementptr inbounds i8, ptr %p2, i64 %index
  IR   %ld2 = load i8, ptr %arrayidx1, align 1
  IR   %cmp3 = icmp eq i8 %ld1, %ld2
No successors
}
```

Mel-Chen committed:
Update llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

Use explicit lambda capture instead of default capture-all

Co-authored-by: Luke Lau <luke_lau@icloud.com>

Mel-Chen committed:
Add TODO for interleaved load

Mel-Chen committed:
Revert "use NCD-based sinking"

This reverts commit bbecac428b70f82a08c3c37e7dca23753d6f823d.

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2547,6 +2547,69 @@ static void licm(VPlan &Plan) {
+        continue;
+      // Skip if the defining block does not dominate the sink block.
+      if (!VPDT.properlyDominates(VPBB, SinkBB))
+        continue;
```
> Also, I’d prefer recipes to be sunk only into exit blocks, rather the latest possible placement (the NCD). 

> The main reason is that we sometimes compute the cost of blocks outside the vectorized loop, for example in isOutsideLoopWorkProfitable. If a recipe that is supposed to be costed is sunk further down to a non-exit block, it’s quite possible that we would miss accounting for that recipe.

That seems like something to fix in isOutsideLoopWorkProfitable, not in licm.

My strong preference would be that we don't restrict it to exit blocks because the middle block can have multiple successors, e.g. see the diagram here: https://llvm.org/docs/Vectorizers.html#id21 we might want to sink it into scalar.ph and not in the path of latch.exit.

But in any case if we really want to restrict sinking to just exit blocks, can you just check `SinkBB->getSinglePredecessor() == LoopRegion` instead of keeping around the PtrSet?

Mel-Chen committed:
rebase  and update

Mel-Chen committed:
early break if there is not only one block use the value.

Co-authored-by: Luke Lau luke@igalia.com

Mel-Chen committed:
remove SmallPtrSet<VPBasicBlock *, 2> DedicatedExits

Mel-Chen commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2533,6 +2533,69 @@ static void licm(VPlan &Plan) {
+
+      VPBasicBlock *SinkBB = nullptr;
+      // TODO: Support sinking when users are in multiple blocks.
+      if (UserBBs.size() == 1) {
```
f58f05be53bf4f71ec546a0ab67539fdc49be87a

lukel97 reviewed (approved):
LGTM, just minor comments. Ran this on llvm-test-suite for rva23u64 and happy to report I didn't see any legacy cost model assertions

fhahn reviewed (commented):
I discovered 2 crashes while testing, added in e34fefdb35aa647a23fdd0240c9c30e6c2f324ed. One was a cost-mismatch, the other due to sinking replicating VPReplciateRecipe after a quick glance. I think we may need to be more careful around those or make sure replicateByVF also runs on the sunk ones

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2545,6 +2545,58 @@ static void licm(VPlan &Plan) {
+      VPBasicBlock *SinkBB = *UserBBs.begin();
+      // Only sink to dedicated exit blocks.
+      if (SinkBB->getSinglePredecessor() != LoopRegion)
+        continue;
```
In #172454 this PR would sink VPInstruction::FirstActiveLanes to the `vector.early.exit` blocks created by VPlanTransforms::handleUncountableEarlyExit . As I can tell each one always has a single successor.

Mel-Chen committed:
updated after #176754

Mel-Chen committed:
Still need to patch planContainsAdditionalSimplifications for cost model
ICE

Mel-Chen committed:
Fix replicateByVF processes the def after use

Mel-Chen committed:
Or another fix way if we don't want to change the traversal way.

Mel-Chen committed:
Replace if-check with assert since we have not support conditinal block
in vectorized loop

Mel-Chen committed:
Remove non-dedicated exits TODO comment

Mel-Chen committed:
Skip all VPReplicateRecipe when licm sinking

Mel-Chen committed:
refactor sinkable recipe checking

Mel-Chen committed:
refine comment of sinkable recipe checking

Mel-Chen committed:
Add store for some test cases

Mel-Chen committed:
restore unrelated test change

Mel-Chen committed:
improve the planContainsAdditionalSimplifications checking

Mel-Chen committed:
move VPReplicateRecipe checking forward

Mel-Chen committed:
add comment

Mel-Chen committed:
update comment

Mel-Chen committed:
update comment

Mel-Chen commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7149,6 +7149,14 @@ static bool planContainsAdditionalSimplifications(VPlan &Plan,
+        // and the VPlan-based cost model.
+        if (auto *Def = dyn_cast<VPSingleDefRecipe>(&R))
+          if (Def->isDefinedOutsideLoopRegions())
+            return true;
```
We can directly check the block of sunk recipe.
284acd4ccdef54fa4502c4ae51d3b54dfc0ba243

Mel-Chen commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2545,6 +2545,60 @@ static void licm(VPlan &Plan) {
+      // TODO: Support sinking VPReplicateRecipe after ensuring replicateByVF
+      // handles sunk recipes correctly.
+      if (isa<VPReplicateRecipe>(Def))
+        continue;
```
d74bc22b804704ca8b9eb96296568777cce582a6

lukel97 reviewed (commented):
Ran this on aarch64 with -mcpu=apple-m1 -O3 -flto on macOS and didn't see anymore crashes, but I couldn't get the assertions to trigger without the `planContainsAdditionalSimplifications` checks. I guess the assertions came from a proprietary workload.

fhahn reviewed (approved):
LGTM, thanks

Mel-Chen committed:
updated comment
